Add Terraform version information to Cloud Run Services

upodroid opened the pull request:
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/5385

<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->




<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
cloudrun: Terraform now adds `client-version` and `client-name` annotations to `google_cloudrun_service`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@rileykarson, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 12 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=188266"

rileykarson commented:
Hmm- I'm seeing a unit test error:

```
# github.com/hashicorp/terraform-provider-google/google
google/resource_cloud_run_service.go:2379:39: undefined: version
```

upodroid commented:
I don't see that error locally.

```
C02S62ZFFVH6:terraform-provider-google mahamed$ git status
On branch master
Your branch is behind 'origin/master' by 20 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   .go-version
        modified:   go.mod
        modified:   go.sum
        modified:   google/resource_cloud_run_service.go

no changes added to commit (use "git add" and/or "git commit -a")
C02S62ZFFVH6:terraform-provider-google mahamed$ git rev-parse HEAD
dd1233ca80489d3c9b332c0e7642c943a12e01d8
C02S62ZFFVH6:terraform-provider-google mahamed$ make testacc TEST=./google TESTARGS='-run=TestAccCloudRunService_cloudRunServiceBasicExample'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudRunService_cloudRunServiceBasicExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudRunService_cloudRunServiceBasicExample
=== PAUSE TestAccCloudRunService_cloudRunServiceBasicExample
=== CONT  TestAccCloudRunService_cloudRunServiceBasicExample
--- PASS: TestAccCloudRunService_cloudRunServiceBasicExample (130.77s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google   135.373s
```

rileykarson commented:
Let me try and rebuild. It looks like your downstream is behind HEAD by a week or two, it could be worth doing a pull on `terraform-provider-google` and regenerating.

/gcbrun

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190012"

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 12 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190013"

upodroid commented:
Rebased to a newer commit.

```
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google   master  5⬇  3+  $    make test
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
go test  -timeout=30s $(go list ./...)
?       github.com/hashicorp/terraform-provider-google  [no test files]
ok      github.com/hashicorp/terraform-provider-google/google   23.715s
?       github.com/hashicorp/terraform-provider-google/scripts/affectedtests    [no test files]
?       github.com/hashicorp/terraform-provider-google/scripts/sidebar  [no test files]
?       github.com/hashicorp/terraform-provider-google/version  [no test files]
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google   master  5⬇  3✎  $    git status
On branch master
Your branch is behind 'origin/master' by 5 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   google/resource_cloud_run_domain_mapping.go
        modified:   google/resource_cloud_run_service.go
        modified:   website/docs/r/cloud_run_service.html.markdown

no changes added to commit (use "git add" and/or "git commit -a")
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google   master  5⬇  3✎  $    git rev-parse HEAD
89008b41015bdd5fa365bd481a8aba55f6ae7d32
 REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google   master  5⬇  3✎  $    make testacc TEST=./google TESTARGS='-run=TestAccCloudRunService_cloudRunServiceBasicExample'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccCloudRunService_cloudRunServiceBasicExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccCloudRunService_cloudRunServiceBasicExample
=== PAUSE TestAccCloudRunService_cloudRunServiceBasicExample
=== CONT  TestAccCloudRunService_cloudRunServiceBasicExample
--- PASS: TestAccCloudRunService_cloudRunServiceBasicExample (45.39s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google   46.436s
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 12 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190024"

rileykarson commented:
Hmm- still seeing an issue, I'll try it again.

/gcbrun

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190416"

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 36 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 12 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190417"

rileykarson commented:
Huh- it seems `goimports` isn't recognising that the `version` package needs to be imported when creating `auto-pr-`s. Want to add a function in one of `provider.go`/`config.go`/`utils.go` that returns the version, and use that instead of referencing `version.ProviderVersion` directly? We can control the imports there manually.

upodroid commented:
I reworked it. Ideally, it would have been `config.providerVersion` but `config` isn't available in decoders :(

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 40 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 41 insertions(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 28 insertions(+))

modular-magician commented:
I have triggered VCR tests based on this PR's diffs. See the results here: "https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=190428"

upodroid commented:
bump and thank you for the OSPB nomination

rileykarson commented:
Hmm- do you mind rebasing? I think the changes in `client.go` may also need to be amended based on some changes there.

I'm not certain how safe I feel this change is to make outside of a major version, as it will cause issues if the `client-name` and `client-version` are already encoded in config. We can consider it in `4.0.0`, though. I'll update the issue.

upodroid commented:
I have been observing the last few releases. No one seems to be bumping this value

https://github.com/GoogleCloudPlatform/magic-modules/blob/7e637acdc0861fe2f5b109e40e3c8a10c05f093e/mmv1/third_party/terraform/version/version.go#L5

so all providers have dev in their user agents

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 28 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 29 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 6 insertions(+))

rileykarson commented:
That value gets set at build time, it's intentionally set to `dev` in code. I checked the most recent `google` release and it gets set correctly:

```
---[ REQUEST ]---------------------------------------
POST /compute/v1/projects/my-project/regions/us-central1/addresses?alt=json HTTP/1.1
Host: compute.googleapis.com
User-Agent: Terraform/0.13.4 (+https://www.terraform.io) Terraform-Plugin-SDK/2.5.0 terraform-provider-google/3.75.0
Content-Length: 52
Content-Type: application/json
Accept-Encoding: gzip

{
 "addressType": "EXTERNAL",
 "name": "address-test-ma"
}
```

rileykarson commented on mmv1/third_party/terraform/utils/config.go.erb:
```diff
@@ -78,6 +83,8 @@ type providerMeta struct {
 	ModuleName string `cty:"module_name"`
 }
 
+var providerVersion string
```
Instead of a global var, let's set this during `providerConfigure` in `provider.go`: https://github.com/borg-land/magic-modules/blob/cloudrun-versioning-annotations/mmv1/third_party/terraform/utils/provider.go.erb#L505

We can share code w/ the useragent creation in that file.

That should also sidestep the tgc failure (config.go is used there, provider.go isn't)

upodroid replied:
Fixed

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 718 insertions(+), 693 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 25 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 718 insertions(+), 693 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 4 files changed, 26 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 718 insertions(+), 693 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 25 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 718 insertions(+), 693 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 26 insertions(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 718 insertions(+), 693 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 26 insertions(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

rileykarson commented:
Hmm- looks like there's a build issue. Line numbers from tpg:

```
go install
# github.com/hashicorp/terraform-provider-google/google
google/provider.go:1242:25: syntax error: unexpected =, expecting comma or }
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 26 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 26 insertions(+), 1 deletion(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

upodroid commented:
Fixed

rileykarson commented on mmv1/third_party/terraform/utils/provider.go.erb:
```diff
@@ -10,6 +10,11 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 
+<% unless version == 'ga' -%>
```
goimports should figure the right thing out w/o this, now that we're in this file - I don't think we need these lines as hints. We've got a third provider variant (alpha / eap) kicking around in a different namespace (and incorrect hints *won't* get figured out right)

upodroid replied:
Long night, eh?

Fixed

rileykarson reviewed (commented):
Oh man, if I end up more than a week behind please feel free to ping the issue. I lose review emails sometimes (like this one)

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 25 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 25 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))

upodroid commented:
Did we miss this for v4?

upodroid commented:
Also, this isn't a breaking change as the annotation key run.googleapis.com is owned by Google and end users shouldn't be modifying it at all. It is a well known convention in Kubernetes not to use annotation prefixes that don't belong to you.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-4798-old..auto-pr-4798) ( 3 files changed, 27 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-4798-old..auto-pr-4798) ( 4 files changed, 49 insertions(+), 23 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-4798-old..auto-pr-4798) ( 1 file changed, 1 insertion(+))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-4798-old..auto-pr-4798) ( 2 files changed, 12 insertions(+))
